deeplearning/ml4pl/poj104/dataset.py

```diff
@@ -30,16 +31,48 @@
 REPO_ROOT = Path(REPO_ROOT)
 
 
-# The vocabulary file used in the dataflow experiments.
-PROGRAML_VOCABULARY = REPO_ROOT / "deeplearning/ml4pl/poj104/programl_vocabulary.csv"
-assert PROGRAML_VOCABULARY.is_file()
+# The vocabulary files used in the dataflow experiments.
+PROGRAML_VOCABULARY = Path("deeplearning/ml4pl/poj104/programl_vocabulary.csv")
+CDFG_VOCABULARY = Path("deeplearning/ml4pl/poj104/cdfg_vocabulary.csv")
+assert PROGRAML_VOCABULARY.is_file(), f"File not found: {PROGRAML_VOCABULARY}"
+assert CDFG_VOCABULARY.is_file(), f"File not found: {CDFG_VOCABULARY}"
 
+# The path of the graph2cdfg binary which converts ProGraML graphs to the CDFG
+# representation.
+#
+# To build this file, clone the ProGraML repo and build
+# //programl/cmd:graph2cdfg:
+#
+#   1.  git clone https://github.com/ChrisCummins/ProGraML.git
+#   2.  cd ProGraML
+#   3.  git checkout 2d93e5e14bf321336f1928d3364e9d7196cee995
+#   4.  bazel build -c opt //programl/cmd:graph2cdfg
+#   5.  cp -v bazel-bin/programl/cmd/graph2cdfg ${THIS_DIR}
+#
+GRAPH2CDFG = Path("deeplearning/ml4pl/poj104/graph2cdfg")
+assert GRAPH2CDFG.is_file(), f"File not found: {GRAPH2CDFG}"
 
-def load(file: str) -> ProgramGraph:
-    """Read a ProgramGraph protocol buffer from file."""
+
+def load(file: str, cdfg: bool = False) -> ProgramGraph:
+    """Read a ProgramGraph protocol buffer from file.
+
+    Args:
+        file: The path of the ProgramGraph protocol buffer to load.
+        cdfg: If true, convert the graph to CDFG during load.
+    """
     graph = ProgramGraph()
     with open(file, 'rb') as f:
-        graph.ParseFromString(f.read())
+        proto = f.read()
+
+    if cdfg:
+        graph2cdfg = subprocess.Popen(
+            [str(GRAPH2CDFG), '--stdin_fmt=pb', '--stdout_fmt=pb'],
+            stdin=subprocess.PIPE,  stdout=subprocess.PIPE
+        )
+        proto, _ = graph2cdfg.communicate(proto)
+        assert not graph2cdfg.returncode, f"CDFG conversion failed: {file}"
```

Comment:
Cool! Thanks a lot for staying in python land with this code base 👍 
It's maybe the second time in my life that I need this 'pipe to subprocess' pattern and nice to see it here so cleanly executed!

Reply:
Easier or everyon! The downside is that the binary I committed won't work for everyone. In *theory* it should work on any vaguely-Ubuntu Linux distro, but if you encounter errors you'll have to build it.